crates/holochain/src/core/workflow/incoming_dht_ops_workflow.rs

```diff
@@ -110,24 +110,54 @@ fn batch_process_entry(
     Ok(())
 }
 
-#[instrument(skip(vault, sys_validation_trigger, ops))]
+#[derive(Default, Clone)]
+pub struct IncomingOpHashes(Arc<parking_lot::Mutex<HashSet<DhtOpHash>>>);
+
+#[instrument(skip(vault, sys_validation_trigger, ops, incoming_op_hashes))]
 pub async fn incoming_dht_ops_workflow(
     vault: &EnvWrite,
+    incoming_op_hashes: Option<&IncomingOpHashes>,
     mut sys_validation_trigger: TriggerSender,
-    ops: Vec<(holo_hash::DhtOpHash, holochain_types::dht_op::DhtOp)>,
+    mut ops: Vec<(holo_hash::DhtOpHash, holochain_types::dht_op::DhtOp)>,
     request_validation_receipt: bool,
 ) -> WorkflowResult<()> {
     let mut filter_ops = Vec::new();
+    let mut hashes_to_remove = if incoming_op_hashes.is_some() {
+        Vec::with_capacity(ops.len())
+    } else {
+        Vec::with_capacity(0)
+    };
+
+    if let Some(incoming_op_hashes) = &incoming_op_hashes {
+        let mut set = incoming_op_hashes.0.lock();
+        let mut o = Vec::with_capacity(ops.len());
```

Comment:
Is this comment I'm adding right?

```suggestion
    // Filter out ops that are already being tracked, so we don't do duplicate work
    if let Some(incoming_op_hashes) = &incoming_op_hashes {
        let mut set = incoming_op_hashes.0.lock();
        let mut o = Vec::with_capacity(ops.len());
```